From b6bc5117ded379348ab368dc8dfe2f77b1586a90 Mon Sep 17 00:00:00 2001
From: pchome <pchome@users.noreply.github.com>
Date: Sun, 24 May 2026 07:11:39 +0300
Subject: [PATCH] build: fix -Wrange-loop-construct warning

---
 source/effect_preprocessor.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/source/effect_preprocessor.cpp b/source/effect_preprocessor.cpp
index 6e09c77..3f48b6e 100644
--- a/source/effect_preprocessor.cpp
+++ b/source/effect_preprocessor.cpp
@@ -165,7 +165,7 @@ std::vector<std::filesystem::path> reshadefx::preprocessor::included_files() con
 {
 	std::vector<std::filesystem::path> files;
 	files.reserve(_file_cache.size());
-	for (const std::pair<std::string, std::string> &cache_entry : _file_cache)
+	for (const std::pair<const std::string, std::string> &cache_entry : _file_cache)
 		files.push_back(std::filesystem::u8path(cache_entry.first));
 	return files;
 }
-- 
2.53.0

