diff --git a/source/effect_codegen_spirv.cpp b/source/effect_codegen_spirv.cpp
index 7b0f538..d978bc9 100644
--- a/source/effect_codegen_spirv.cpp
+++ b/source/effect_codegen_spirv.cpp
@@ -373,8 +373,51 @@ private:
 
 	std::string finalize_code() const override
 	{
-		// There is no high-level text representation
-		return std::string();
+		std::basic_string<char> spirv;
+		finalize_header_section(spirv);
+
+		// All entry point declarations
+		for (const spirv_instruction &inst : _entries.instructions)
+			inst.write(spirv);
+
+		// All execution mode declarations
+		for (const spirv_instruction &inst : _execution_modes.instructions)
+			inst.write(spirv);
+
+		finalize_debug_info_section(spirv);
+
+		for (const spirv_instruction &inst : _debug_b.instructions)
+			inst.write(spirv);
+
+		// All annotation instructions
+		for (const spirv_instruction &inst : _annotations.instructions)
+			inst.write(spirv);
+
+		finalize_type_and_constants_section(spirv);
+
+		for (const spirv_instruction &inst : _variables.instructions)
+			inst.write(spirv);
+
+		// All function definitions
+		for (const function_blocks &func : _functions_blocks)
+		{
+			if (func.definition.instructions.empty())
+				continue;
+
+			for (const spirv_instruction &inst : func.declaration.instructions)
+				inst.write(spirv);
+
+			// Grab first label and move it in front of variable declarations
+			func.definition.instructions.front().write(spirv);
+			assert(func.definition.instructions.front().op == spv::OpLabel);
+
+			for (const spirv_instruction &inst : func.variables.instructions)
+				inst.write(spirv);
+			for (auto inst_it = func.definition.instructions.begin() + 1; inst_it != func.definition.instructions.end(); ++inst_it)
+				inst_it->write(spirv);
+		}
+
+		return spirv;
 	}
 	bool assemble_code_for_entry_point(const std::string &entry_point_name, std::string &spirv, std::string &, std::string &) const override
 	{
